src/zcap.rs

```diff
@@ -0,0 +1,580 @@
+#[serde(rename_all = "camelCase")]
+pub struct Invocation<S = DefaultProps<String>> {
+    #[serde(rename = "@context")]
+    pub context: Contexts,
+    pub id: URI,
+    #[serde(flatten)]
+    pub property_set: S,
+    // This field is populated only when using
+    // embedded proofs such as LD-PROOF
+    //   https://w3c-ccg.github.io/ld-proofs/
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub proof: Option<Proof>,
+}
+
+impl<S> Invocation<S> {
+    pub fn new(id: URI, property_set: S) -> Self {
+        Self {
+            context: Contexts::default(),
+            id,
+            proof: None,
+            property_set,
+        }
+    }
+}
+
+pub trait VerifyAttributes<I> {
+    type Err: std::string::ToString;
+    // Fn to be implemented for invocations to be verified against delegations
+    // If the property set contains e.g. Actions or Caveat-related fields, they should be checked here
+    fn verify_attributes(&self, invocation: &I) -> Result<(), Self::Err> {
+        Ok(())
+    }
+}
+
+impl<S> Invocation<S>
+where
+    S: Serialize + Send + Sync + Clone,
+{
+    pub async fn verify<C, P>(
+        &self,
+        _options: Option<LinkedDataProofOptions>,
+        resolver: &dyn DIDResolver,
+        // TODO make this a list for delegation chains
+        target_capability: &Delegation<C, P>,
+    ) -> VerificationResult
+    where
+        C: Serialize + Send + Sync + Clone,
+        P: Serialize + Send + Sync + Clone,
+        Delegation<C, P>: VerifyAttributes<Self>,
+    {
+        match &self.proof {
+            None => VerificationResult::error("No applicable proof"),
+            Some(proof) => {
+                let mut result = proof.verify(self, resolver).await;
+                match (
+                    // get cap id from proof extra properties
+                    proof
+                        .property_set
+                        .as_ref()
+                        .and_then(|ps| ps.get("capability").map(|s| s.clone()))
+                        .and_then(|v| match v {
+                            Value::String(id) => Some(id),
+                            _ => None,
+                        }),
+                    &target_capability.id,
+                ) {
+                    (Some(ref id), URI::String(ref t_id)) => {
+                        // ensure proof target cap ID and given
+                        if id != t_id {
+                            result
+                                .errors
+                                .push("Target Capability IDs dont match".into())
+                        };
+                    }
+                    _ => result
+                        .errors
+                        .push("Missing proof target capability ID".into()),
+                };
+                match (&target_capability.invoker, &proof.verification_method) {
+                    // if there are invokers listed in the target, ensure the inoker here is the right one
```

Comment:
Update typo, and update meaning as mentioned in https://github.com/spruceid/ssi/pull/189#discussion_r661670808, and to note about using target_capability's capabilityDelegation verification methods
```suggestion
                    // Ensure the proof's verification method is authorized as an invoker. TODO: also allow target_capability's capabilityDelegation verification methods.
```

Reply:
[made an issue about this here](https://github.com/spruceid/ssi/issues/220)